### PR TITLE
utils and libraries: address check in update-translations.py

### PR DESCRIPTION
**Pull request by abskj**
fix for issue #13363

### CODE REVIEW

File: contrib/devtools/update-translations.py [hunk 1]
```diff
@@ -122,6 +122,12 @@ def escape_cdata(text):
     text = text.replace('"', '&quot;')
     return text
 
+def check_for_address(s):
+    #checks each message for bitcoin addresses
+    address_rgx='(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,34}'
```
Comment (MarcoFalke) [inline contrib/devtools/update-translations.py]: nit: Can make an upper case constant with re.compile

File: contrib/devtools/update-translations.py [hunk 2]
```diff
@@ -122,6 +122,12 @@ def escape_cdata(text):
...
+    return bool(re.search(address_rgx,s))
```
Comment (MarcoFalke) [inline contrib/devtools/update-translations.py]: Probably fine to just inline (using the compiled re constant), since it is only used once

**Comment by Empact** [comment]
See also #13374

File: contrib/devtools/update-translations.py [hunk 3]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
```
Comment (giancarloGiuffra) [inline contrib/devtools/update-translations.py]: hi, delete the whitespace to make the build pass. The linter is failing.

Comment (giancarloGiuffra) [reply contrib/devtools/update-translations.py]: it's in line 164

Comment (ccdle12) [reply contrib/devtools/update-translations.py]: Can also install ```clang-format``` and run https://github.com/bitcoin/bitcoin/tree/master/contrib/devtools#clang-format-diffpy to fix any whitespaces before pushing

Comment (undercoverGod) [reply contrib/devtools/update-translations.py]: done..thank you

File: contrib/devtools/update-translations.py [hunk 4]
```diff
@@ -30,6 +30,8 @@
 LOCALE_DIR = 'src/qt/locale'
 # Minimum number of messages for translation to be considered at all
 MIN_NUM_MESSAGES = 10
+#regex for address
```
Comment (promag) [inline contrib/devtools/update-translations.py]: Nit, `# Regular expression for addresses`

File: contrib/devtools/update-translations.py [hunk 5]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
+                    #check if btc address is included in translation
```
Comment (promag) [inline contrib/devtools/update-translations.py]: Nit `# Check ...`

File: contrib/devtools/update-translations.py [hunk 6]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
+                    # Check if btc address is included in translation
+                    if re.search(ADDRESS_RE,translation):
```
Comment (promag) [inline contrib/devtools/update-translations.py]: nit, space after `,`.

File: contrib/devtools/update-translations.py [hunk 7]
```diff
@@ -178,6 +185,10 @@ def postprocess_translations(reduce_diff_hacks=False):
                 if translation_node.get('type') == 'unfinished':
                     context.remove(message)
 
+                if translation_node.get('type') == 'malicious' :
```
Comment (promag) [inline contrib/devtools/update-translations.py]: nit, remove space before `:`

File: contrib/devtools/update-translations.py [hunk 8]
```diff
@@ -30,6 +30,8 @@
 LOCALE_DIR = 'src/qt/locale'
 # Minimum number of messages for translation to be considered at all
 MIN_NUM_MESSAGES = 10
+# Regular expression for addresses
+ADDRESS_RE = re.compile('(bc1|[13])[a-zA-HJ-NP-Z0-9]{25,34}')
```
Comment (promag) [inline contrib/devtools/update-translations.py]: How about multisig?

File: contrib/devtools/update-translations.py (hunk 6 above)
Comment (promag) [reply contrib/devtools/update-translations.py]: Maybe remove whitespaces from `translation` before searching for addresses?

**Review by promag — COMMENTED** [review]
Concept ACK.

File: contrib/devtools/update-translations.py (hunk 8 above)
Comment (undercoverGod) [reply contrib/devtools/update-translations.py]: I thought the address format is same for multisig. Could you provide some info about it if I am mistaken.

Comment (promag) [reply contrib/devtools/update-translations.py]: Sorry it's already covered. Maybe change to `(1|3|bc1)`?

Comment (undercoverGod) [reply contrib/devtools/update-translations.py]: done

**Review by giancarloGiuffra — APPROVED** [review]
ACK, tested with P2PKH, P2SH and Bech32 addresses.

File: contrib/devtools/update-translations.py [hunk 9]
```diff
@@ -178,6 +185,10 @@ def postprocess_translations(reduce_diff_hacks=False):
                 if translation_node.get('type') == 'unfinished':
                     context.remove(message)
 
+                if translation_node.get('type') == 'malicious':
+                    print('WARNING...address found in translation.Message will be deleted')
```
Comment (Empact) [inline contrib/devtools/update-translations.py]: Would be good to output the filename here for specificity, as with line 173.

File: contrib/devtools/update-translations.py [hunk 10]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
+                    # Check if btc address is included in translation after removing whitespaces
+                    if re.search(ADDRESS_RE, translation.replace(' ', '')):
+                        translation_node.set('type', 'malicious')
+                        break
                     errors = []
                     valid = check_format_specifiers(source, translation, errors, numerus)
```
Comment (Empact) [inline contrib/devtools/update-translations.py]: Independent of this PR, but this bit could be simplified a fair bit:
* Return error from `check_format_specifiers`, which generates at most 1 error
* Combine loop and if not valid below

File: contrib/devtools/update-translations.py [hunk 11]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
+                    # Check if btc address is included in translation after removing whitespaces
+                    if re.search(ADDRESS_RE, translation.replace(' ', '')):
+                        translation_node.set('type', 'malicious')
+                        break
```
Comment (Empact) [inline contrib/devtools/update-translations.py]: nit: I still like the idea of using errors for this.

File: contrib/devtools/update-translations.py [hunk 12]
```diff
@@ -159,6 +161,11 @@ def postprocess_translations(reduce_diff_hacks=False):
                 for translation in translations:
                     if translation is None:
                         continue
+
+                    # Check if btc address is included in translation after removing whitespaces
+                    if re.search(ADDRESS_RE, translation.replace(' ', '')):
```
Comment (promag) [inline contrib/devtools/update-translations.py]: To remove all whitespaces:
```py
translation.translate( { ord(c):None for c in ' \n\t\r' } )
```
ot
```py
''.join(translation.split())
```

**Review by promag — COMMENTED** [review]
Please squash.

**Comment by promag** [comment]
Just realised that this is duplicate of #13374. @kaplanmaxe maybe incorporate above suggestions in yours?

File: contrib/devtools/update-translations.py (hunk 11 above)
Comment (undercoverGod) [reply contrib/devtools/update-translations.py]: Yes..realized now. Amending.

File: contrib/devtools/update-translations.py (hunk 12 above)
Comment (undercoverGod) [reply contrib/devtools/update-translations.py]: done..the implementation is changed in accordance with @Empact 's suggestion

File: contrib/devtools/update-translations.py [hunk 13]
```diff
@@ -80,6 +82,10 @@ def sanitize_string(s):
     return s.replace('\n',' ')
 
 def check_format_specifiers(source, translation, errors, numerus):
+    # Check for btc address in translation after removing whitespaces
+    if re.search(ADDRESS_RE, ''.join(translation.split())):
+        errors.append("WARNING...Address found in translation for '%s': '%s'\nMessage will be deleted" % (sanitize_string(source), sanitize_string(translation)))
```
Comment (laanwj) [inline contrib/devtools/update-translations.py]: There's not need to add the "\nMessage will be deleted" here, it's the case for all errors so doesn't need mentioning.

**Comment by laanwj** [comment]
Tested ACK 85f0135eaefe3d9f696689a7e83606c579da40a8

**Comment by laanwj** [comment]
Closing, as  #13374 was merged.

**Pull request closed without merging**